### ISSUE TITLE
Alacritty --title seems to not set Wayland title

### ISSUE THREAD

**Issue by bemeurer** [jwilm/alacritty#2104]
Running the latest master under Sway/Wayland and Gentoo it seems like running Alacritty with `-t foo` doesn't correctly set the title on Wayland. For example, I have the following line in my Sway config

```
set $menu gnome-terminal --profile aopen -t aopen --class aopen -- ~/bin/aopen
for_window [title="aopen"] floating enable, border none
```
This works file, here's an example

![image](https://user-images.githubusercontent.com/7243783/52599035-59a54780-2e0c-11e9-84bd-cb0c536ba800.png)

Changing that line to use Alacritty, looks like this

```
set $menu ~/src/alacritty/target/release/alacritty --title aopen --class aopen -e ~/bin/aopen
for_window [title="aopen"] floating enable, border none
```
And yields the following, incorrect, result
![image](https://user-images.githubusercontent.com/7243783/52599157-b7399400-2e0c-11e9-8814-43c25538427f.png)

**Comment by chrisduerr** [comment]
Does window class work correctly? I believe that has recently been added and should give correct results.

Window title support might be missing in winit, though I'm not sure.

Just for reference, the X.Org output seems correct, indicating this might be a winit issue:
```
WM_CLASS(STRING) = "alacritty-class", "Alacritty"
_NET_WM_NAME(UTF8_STRING) = "alacritty-title"
WM_NAME(STRING) = "alacritty-title"
```

**Comment by chrisduerr** [comment]
For reference, I found this issue mentioned in the winit tracker here:
https://github.com/tomaka/winit/issues/306#issuecomment-339308555

**Comment by bemeurer** [comment]
So, for a working window this is what one portion of `swaymsg -t get_tree` looks like:
```
 "window_properties": {
                 "class": "Firefox",
                 "instance": "Navigator",
                 "title": "Alacritty --title seems to not set Wayland title · Issue #2104 · jwilm\/alacritty - Mozilla Firefox",
                 "transient_for": null,
                 "window_role": "browser"
               },
```
That same section, however, is not set for Alacritty.

**Comment by bemeurer** [comment]
I found a workaround! Apparently alacritty does set `app_id` correctly! So the following works:
```
set $menu ~/src/alacritty/target/release/alacritty --title aopen --class aopen -e ~/bin/aopen
for_window [app_id="aopen"] floating enable, border none
```
![image](https://user-images.githubusercontent.com/7243783/52600284-f3babf00-2e0f-11e9-98cf-6b277d9d6155.png)

**Comment by trimental** [comment]
https://github.com/tomaka/winit/pull/770 Does this fix that?

**Comment by rkanati** [comment]
If it helps, I was having trouble with exactly this problem (before I found `app_id`), and I found that if I run `alacritty --title oops` from an existing instance, _that_ instance gets its title changed, and the new instance opens with a regular/default title.

**Comment by chrisduerr** [comment]
@rkanati I'd assume this is with the current version of Alacritty?

Can you confirm whether the PR trimental proposed works?

**Comment by rkanati** [comment]
Turns out I was wrong about the first instance changing title, my bad.

Trying to check trimental's PR - figuring out how to change a dependency of a dependency right now.

**Comment by chrisduerr** [comment]
@rkanati You'll have to clone glutin inside Alacritty, then change glutin in our Cargo.toml to use `glutin = { path = "./glutin", ... }` and then you can just change glutin's winit to `winit = { git = "https://...", ... }`.

That's at least how I usually do it. You can probably also use the cargo patch directive, but I wouldn't know how to use it without looking it up.

**Comment by rkanati** [comment]
That's exactly what I tried... but alacritty needs `glutin = "0.12`, and that version is mysteriously absent from tomaka/winit 

Now seeing if Cargo has a source-tarball-like feature...

**Comment by chrisduerr** [comment]
Glutin depends on winit, not the other way around.

That's why you'd need to clone glutin, edit Alacritty's Cargo.toml to link to the cloned glutin and then edit glutin's Cargo.toml to link to the PR's winit.

**Comment by rkanati** [comment]
Sorry, I meant glutin 0.12 was absent from its own git repo.

It's okay though, I think I've got it. Give me a few minutes.

**Comment by chrisduerr** [comment]
You can just clone the latest master. As long as there were no breaking updates Alacritty should still work great.

**Comment by rkanati** [comment]
Okay, unless I'm missing something, alacritty won't build against glutin:master. If trimental's patch fixes the problem, it doesn't seem we can benefit from it unless we can bump our glutin dep out of the stone age :disappointed:

**Comment by chrisduerr** [comment]
>  it doesn't seem we can benefit from it unless we can bump our glutin dep out of the stone age

Alacritty is using version 0.19, which is the latest released version of glutin.

> Okay, unless I'm missing something, alacritty won't build against `glutin:master`

It definitely looks to me like you might be missing something, since Alacritty builds just fine against the latest glutin for me.

The following diff will allow you to use a local copy of glutin in the same directory as `src` and all the other Alacritty files:

```diff
diff --git a/Cargo.toml b/Cargo.toml
index ce22a3d..67d0bf7 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -42,7 +42,7 @@ clap = "2"
 fnv = "1"
 unicode-width = "0.1"
 arraydeque = "0.4"
-glutin = { version = "0.19", features = ["icon_loading"] }
+glutin = { path = "./glutin", features = ["icon_loading"] }
 env_logger = "0.6.0"
 base64 = "0.10.0"
 static_assertions = "0.3.0"
```

The patch necessary to build glutin against trimental's fork looks like this:

```diff
diff --git a/Cargo.toml b/Cargo.toml
index 131f12d..5178902 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -21,7 +21,7 @@ serde = ["winit/serde"]
 lazy_static = "1"
 libc = "0.2"
 shared_library = "0.1.0"
-winit = "0.18.0"
+winit = { git = "https://github.com/trimental/winit", branch = "fix-with-title-wayland" }
 
 [build-dependencies]
 gl_generator = "0.10"
```

Hopefully this should help you, or anyone else interested in testing this, with setting everything up. All that is required should be to edit these two files and clone the glutin repository. Though if this is too difficult, setting up the cargo patch shouldn't require any changes to glutin.

**Comment by rkanati** [comment]
Yeah, I know all that - github was just messing with me, sorry. Let me try again with an not-broken version.

**Comment by rkanati** [comment]
Finally! And it looks good to me - title is set in new instances as expected.

**Comment by chrisduerr** [comment]
> Thanks for your help, @chrisduerr

Thank *you* for testing! And thank you @trimental for fixing this so quickly!

**Comment by nostdm** [comment]
Not sure if the fix has propagated all the way, but currently running `alacritty --title=foo` creates a window with a blank/null title in sway (built alacritty from the current master).

**Comment by chrisduerr** [comment]
@nostdm Which Alacritty commit specifically are you running? Is the glutin version bump already included?

**Comment by nostdm** [comment]
I built the AUR `alacritty-git`package which at the time was at commit `b1032bcc6b79135f87f327548e43563da05657fb`.
Gopkg.toml says it's on version 0.19:
```
glutin = { version = "0.19", features = ["icon_loading"] }
```

**Comment by chrisduerr** [comment]
That commit is three commits behind. I'd appreciate it if you could update and test again.

**Comment by trimental** [comment]
@nostdm this would be typical behaviour if https://github.com/tomaka/winit/commit/f0ce5b0c8d418a3bb5efb8131f107fe28ab5af23 was not included in the build as it fixes setting the title when the window is built however setting the title after the window is built was already working. Therefore com setting the title would work but the title immediately set by alacritty would not.

**Comment by rkanati** [comment]
@nostdm @trimental A quick check of glutin 0.19's `Cargo.toml` shows it depends on winit 0.18, which doesn't include trimental's fix... because glutin 0.19 was released in November. Dang.

A new release of glutin sure would be nice.

**Comment by nostdm** [comment]
@chrisduerr I just compiled from HEAD and everything seems to work now. Thanks a lot! :champagne:

**Comment by chrisduerr** [comment]
@rkanati Glutin 0.20 has actually been released and it does include the desired functionality.

Thanks @nostdm for the confirmation. I've created #2180 to update the change log entry to show this change.

**Comment by bemeurer** [comment]
Awesome! Thanks everyone for all the hard work!
